packages/astro/src/vite-plugin-astro/compile.ts

```diff
@@ -0,0 +1,147 @@
+import { fileURLToPath } from 'url';
+import { ESBuildTransformResult, transformWithEsbuild } from 'vite';
+import { AstroConfig } from '../@types/astro';
+import { cachedCompilation, CompileProps, CompileResult } from '../core/compile/index.js';
+import { LogOptions } from '../core/logger/core.js';
+import { getFileInfo } from '../vite-plugin-utils/index.js';
+
+interface CachedFullCompilation {
+	compileProps: CompileProps;
+	rawId: string;
+	logging: LogOptions;
+}
+
+interface FullCompileResult extends Omit<CompileResult, 'map'> {
+	map: ESBuildTransformResult['map'];
+}
+
+interface EnhanceCompilerErrorOptions {
+	err: Error;
+	id: string;
+	source: string;
+	config: AstroConfig;
+	logging: LogOptions;
+}
+
+const FRONTMATTER_PARSE_REGEXP = /^\-\-\-(.*)^\-\-\-/ms;
+
+export async function cachedFullCompilation({
+	compileProps,
+	rawId,
+	logging,
+}: CachedFullCompilation): Promise<FullCompileResult> {
+	let transformResult: CompileResult;
+	let esbuildResult: ESBuildTransformResult;
+
+	try {
+		transformResult = await cachedCompilation(compileProps);
+		// Compile all TypeScript to JavaScript.
+		// Also, catches invalid JS/TS in the compiled output before returning.
+		esbuildResult = await transformWithEsbuild(transformResult.code, rawId, {
+			loader: 'ts',
+			target: 'esnext',
```

Comment:
Use `esnext` as we don't need down-levelling (same behaviour as `esbuild.transform` and how Vite works currently).